elasticsearch/src/main/scala/com/snowplowanalytics/snowplow/loader/clients/ElasticsearchSenderHTTP.scala

```diff
@@ -91,21 +94,22 @@ class ElasticsearchSenderHTTP(
 
     val newFailures: List[EmitterInput] = if (actions.nonEmpty) {
       futureToTask(client.execute(bulk(actions)))
-        // we rety with linear backoff if an exception happened
+        // we retry with linear back-off if an exception happened
         .retry(delays, exPredicate(connectionAttemptStartTime))
-        .map { bulkResponse =>
-          bulkResponse.items.zip(records)
+        .map {
+          case Right(bulkResponseResponse: RequestSuccess[BulkResponse]) => bulkResponseResponse.result.items.zip(records)
             .map { case (bulkResponseItem, record) =>
               handleResponse(bulkResponseItem.error.map(_.reason), record)
-            }.flatten
+            }.flatten.toList
+          case Left(requestFailure: RequestFailure) => List((requestFailure.error.reason, requestFailure.error.reason.failureNel)) // the bulk request failed
```

Comment:
It seems the api changed and now we directly know if the indexing failed from the response of `client.execute` so we should act accordingly.

Reply:
HTTP operations now return Either[RequestFailure, RequestSuccess[U]] for better error handling. Each of RequestFailure and RequestSuccess contain the http status code, the full json response body, and any http headers, in addition to either the error details or the request resposne type.

Reply:
Changelog doesn't really help to decide: 
`
HTTP operations now return Either[RequestFailure, RequestSuccess[U]] for better error handling. Each of RequestFailure and RequestSuccess contain the http status code, the full json response body, and any http headers, in addition to either the error details or the request resposne type.
`

Reply:
I have done a test. this is my code: 
```scala
client.execute {
    bulk(
      indexInto("myindex" / "mytype").fields("country" -> "Mongolia", "capital" -> "Ulaanbaatar"),
      indexInto("" / "mytype").fields("country" -> "Namibia", "capital" -> "Windhoek")
    )
  }.map {
    case Right(bulkResponseResponse: RequestSuccess[BulkResponse]) =>
      print(bulkResponseResponse.status)
      print(bulkResponseResponse.body)
      print(bulkResponseResponse.result)
    case Left(requestFailure: RequestFailure) =>
      print(requestFailure.error.`type`)
      print(requestFailure.error.reason)
  }
```
The first would a success and the second is a failure. So match ended up in `Right` because Bulk API returned 200.
```
bulkResponseResponse.body is
```
{
  "took": 3,
  "errors": true,
  "items": [
    {
      "index": {
        "_index": "myindex",
        "_type": "mytype",
        "_id": "5YwYKGQBT2drkLy9HHmt",
        "_version": 1,
        "result": "created",
        "_shards": {
          "total": 2,
          "successful": 1,
          "failed": 0
        },
        "_seq_no": 0,
        "_primary_term": 1,
        "status": 201
      }
    },
    {
      "index": {
        "_index": "",
        "_type": "mytype",
        "_id": null,
        "status": 500,
        "error": {
          "type": "string_index_out_of_bounds_exception",
          "reason": "String index out of range: 0"
        }
      }
    }
  ]
}
```
So the same as before we should collect the errors. because `bulkResponseResponse.result` is: 
```scala
BulkResponse(3,true,List(BulkResponseItems(Some(BulkResponseItem(0,5YwYKGQBT2drkLy9HHmt,myindex,mytype,1,false,false,false,created,201,None,Some(Shards(2,0,1)))),None,None), BulkResponseItems(Some(BulkResponseItem(0,null,,mytype,0,false,false,false,null,500,Some(BulkError(string_index_out_of_bounds_exception,String index out of range: 0,null,0,null)),None)),None,None)))```

what do you think?

Reply:
yup, agreed thanks for the investigation